src/Grids/finitedifference.jl

```diff
@@ -135,6 +73,136 @@ function _FiniteDifferenceGrid(topology::Topologies.IntervalTopology)
     )
 end
 
+# called by the FiniteDifferenceGrid constructor, and the ExtrudedFiniteDifferenceGrid constructor with Hypsography
+function fd_geometry_data(
+    face_coordinates::DataLayouts.AbstractData{Geometry.ZPoint{FT}};
+    periodic,
+) where {FT}
+    CT = Geometry.ZPoint{FT}
+    AIdx = (3,)
+    LG = Geometry.LocalGeometry{AIdx, CT, FT, SMatrix{1, 1, FT, 1}}
+    (Ni, Nj, Nk, Nv, Nh) = size(face_coordinates)
+    Nv_face = Nv - periodic
+    Nv_cent = Nv - 1
+
+    center_local_geometry =
+        similar(face_coordinates, LG, (Ni, Nj, Nk, Nv_cent, Nh))
+    face_local_geometry =
+        similar(face_coordinates, LG, (Ni, Nj, Nk, Nv_face, Nh))
+
+    for h in 1:Nh, k in 1:Nk, j in 1:Nj, i in 1:Ni
+
+        for v in 1:Nv_cent
+            # centers
+            coord⁻ = Geometry.component(
+                face_coordinates[CartesianIndex(i, j, k, v, h)],
+                1,
```

Comment:
```suggestion
function fd_geometry_data(
    face_coordinates::DataLayouts.AbstractData{Geometry.ZPoint{FT}};
    periodic,
) where {FT}
    CT = Geometry.ZPoint{FT}
    AIdx = (3,)
    LG = Geometry.LocalGeometry{AIdx, CT, FT, SMatrix{1, 1, FT, 1}}
    (Ni, Nj, Nk, Nv, Nh) = size(face_coordinates)
    Nv_face = Nv - periodic
    Nv_cent = Nv - 1

    center_local_geometry =
        similar(face_coordinates, LG, (Ni, Nj, Nk, Nv_cent, Nh))
    face_local_geometry =
        similar(face_coordinates, LG, (Ni, Nj, Nk, Nv_face, Nh))

    c1(args...) =
        Geometry.component(face_coordinates[CartesianIndex(args...)], 1)

    for h in 1:Nh, k in 1:Nk, j in 1:Nj, i in 1:Ni

        for v in 1:Nv_cent
            # centers
            coord⁻ = c1(i, j, k, v, h)
            coord⁺ = c1(i, j, k, v + 1, h)
            # use a "discrete Jacobian"
            coord = (coord⁺ + coord⁻) / 2
            Δcoord = coord⁺ - coord⁻
            J = Δcoord
            WJ = Δcoord
            ∂x∂ξ = SMatrix{1, 1}(J)
            center_local_geometry[CartesianIndex(i, j, k, v, h)] =
                Geometry.LocalGeometry(
                    CT(coord),
                    J,
                    WJ,
                    Geometry.AxisTensor(
                        (
                            Geometry.LocalAxis{AIdx}(),
                            Geometry.CovariantAxis{AIdx}(),
                        ),
                        ∂x∂ξ,
                    ),
                )
        end

        for v in 1:Nv_face
            coord = c1(i, j, k, v, h)
            if v == 1
                # bottom face
                if periodic
                    Δcoord⁺ = c1(i, j, k, 2, h) - c1(i, j, k, 1, h)
                    Δcoord⁻ = c1(i, j, k, Nv, h) - c1(i, j, k, Nv - 1, h)
                    J = (Δcoord⁺ + Δcoord⁻) / 2
                    WJ = J
                else
                    coord⁺ = c1(i, j, k, 2, h)
                    J = coord⁺ - coord
                    WJ = J / 2
                end
            elseif v == Nv_cent + 1
                @assert !periodic
                # top face
                coord⁻ = c1(i, j, k, Nv - 1, h)
                J = coord - coord⁻
                WJ = J / 2
            else
                coord⁺ = c1(i, j, k, v + 1, h)
                coord⁻ = c1(i, j, k, v - 1, h)
                J = (coord⁺ - coord⁻) / 2
                WJ = J
            end
            ∂x∂ξ = SMatrix{1, 1}(J)
            face_local_geometry[CartesianIndex(i, j, k, v, h)] =
                Geometry.LocalGeometry(
                    CT(coord),
                    J,
                    WJ,
                    Geometry.AxisTensor(
                        (
                            Geometry.LocalAxis{AIdx}(),
                            Geometry.CovariantAxis{AIdx}(),
                        ),
                        ∂x∂ξ,
                    ),
                )
        end
    end
    return (center_local_geometry, face_local_geometry)
end
```
This might be a good place for a closure, just to reduce the code size (check my work, though!)

Reply:
Committed - Indices LGTM (will rebase if it passes all checks.) cc @Sbozzolo